### PR TITLE
Audit log json envelope

### PR DESCRIPTION
**Pull request by UgnineSirdis**
### Changelog entry 

Support the configuration setting to wrap log records into a fixed json envelope, based on the log record nature. To be used when a single-stream unified logs delivery system is in place (like stderr in k8s installations), with all logs streamed there (including audit), and which needs to have some environment-specific information in the message to decide which system to forward a particular record to.

As the first application is for audit logs, the first implementation is also for the audit logs only.

The function is agnostic to the audit logs format, as addresses not the audit system requirements, but the logs delivery and routing system ones.

### Changelog category 

* New feature

### Additional information

...

### CODE REVIEW

File: ydb/core/log_backend/log_backend.cpp [hunk 1]
```diff
@@ -127,36 +170,46 @@ TAutoPtr<TLogBackend> CreateAuditLogUnifiedAgentBackend(
...
-        auto logBackend = CreateAuditLogUnifiedAgentBackend(runConfig, counters);
-        if (logBackend) {
-            auto format = runConfig.AppConfig.GetAuditConfig().GetUnifiedAgentBackend().GetFormat();
-            logBackends[format].push_back(std::move(logBackend));
+        if (auditConfig.HasFileBackend()) {
+            auto logBackend = MaybeWrapWithJsonEnvelope(CreateAuditLogFileBackend(runConfig), auditConfig.GetFileBackend().GetLogJsonEnvelope());
+            if (logBackend) {
+                auto format = auditConfig.GetFileBackend().GetFormat();
+                logBackends[format].push_back(std::move(logBackend));
+            }
```
Comment (ijon) [inline ydb/core/log_backend/log_backend.cpp:199]: `MaybeWrapWithJsonEnvelope()` stands in a way here a little. I would have moved it to the very end, to `logBackends[format].push_back(...)` line -- when the main logic is done. Also one less check to make in `MaybeWrapWithJsonEnvelope()`.

File: ydb/core/log_backend/json_envelope.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        {}
+
+        bool Parse(const TString& replace);
+        void Apply(const TStringBuf& message);
+    };
+
+private:
+    TString TemplateString;
+    NJson::TJsonValue Value;
+    std::vector<TReplace> Replaces;
```
Comment (ijon) [inline ydb/core/log_backend/json_envelope.h:45]: Is there a reason to support multiple placeholder replacements?
Enveloping in general assumes wrapping around a single payload. And single replacement would have made template rendering code much simpler.

File: ydb/core/log_backend/log_backend.cpp [hunk 3]
```diff
@@ -1,9 +1,52 @@
...
+public:
+    TLogBackendWithJsonEnvelope(const TString& jsonEnvelope, THolder<TLogBackend> logBackend)
+        : JsonEnvelope(jsonEnvelope)
+        , LogBackend(std::move(logBackend))
+    {}
+
+    void WriteData(const TLogRecord& rec) override {
+        TString data;
+        TLogRecord record = rec;
+        with_lock (Mutex) {
```
[the remark this answers is not in the document]
Comment (UgnineSirdis) [reply ydb/core/log_backend/log_backend.cpp:20]: I saw other implementations for log backends and they imply multithreaded usage. For example, file backend: https://github.com/ydb-platform/ydb/blob/main/library/cpp/logger/file.cpp#L19

File: ydb/core/log_backend/json_envelope.h (hunk 2 above)
Comment (UgnineSirdis) [reply ydb/core/log_backend/json_envelope.h:45]: No, there is no explicit requirements for such replacement. I can rewrite the code and make it simpler.

File: ydb/core/log_backend/log_backend.cpp [hunk 4]
```diff
@@ -1,9 +1,52 @@
...
 #include <ydb/core/base/counters.h>
 
+#include <util/system/mutex.h>
+
 namespace NKikimr {
 
+class TLogBackendWithJsonEnvelope : public TLogBackend {
+public:
+    TLogBackendWithJsonEnvelope(const TString& jsonEnvelope, THolder<TLogBackend> logBackend)
+        : JsonEnvelope(jsonEnvelope)
```
[the remark this answers is not in the document]
Comment (UgnineSirdis) [reply ydb/core/log_backend/log_backend.cpp:13]: It is allowed to raise exception on misconfiguration in kikime_services_initializers. For example, here: https://github.com/ydb-platform/ydb/blob/main/ydb/core/driver_lib/run/kikimr_services_initializers.cpp#L449

File: ydb/core/log_backend/log_backend.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ijon) [reply ydb/core/log_backend/log_backend.cpp:20]: Mutex in the file backend serializes writes to the single underlying file.
But mutex here, is it strictly necessary? (Also would it be reasonable to have 2 mutexes on a path to writing a single log line in the case of json-enveloping backend over file backend?)

What actions/accesses serializes this mutex? -- the answer is, I suppose: it serializes updates to the single `TJsonEnvelope` member `Value`. 

Can I suggest removing the `Value` member entirely and wrap every log write into a new TJsonValue thus making `TJsonEnvelope` into a pure transformer (without any state updates)? Would it be reasonable?

File: ydb/core/log_backend/log_backend.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (UgnineSirdis) [reply ydb/core/log_backend/log_backend.cpp:13]: Yes, I agree with you that this approach should be improved.
But, by the way, "Attempt to write non utf-8 string" to audit log is wrapped with excpetion handler. In this case we will write message to usual debug log.

**Review by ijon — APPROVED** [review]

**Pull request merged**